### PR TITLE
Restore atom environment when adding project folders to a fresh window

### LINKED ISSUE #10044: Quitting and re-opening Atom does not save the same tabs I had open
Whenever I quit Atom and re-open it, the same 3 tabs from something I was working on 6 months ago always open and _not_ the thing I was just working on.

I have uninstalled and installed the `save-session` package multiple times, no luck.

This is version 1.3.1.

I'm guessing some file somewhere has these 3 files of mine saved and this file just never gets updated?

### LINKED ISSUE #11293: Unsaved modification will be lost when I close multiple windows
### Prerequisites
- [x] Can you reproduce the problem in [safe mode](http://flight-manual.atom.io/hacking-atom/sections/debugging/#check-if-the-problem-shows-up-in-safe-mode)? 
- [x] Are you running the [latest version of Atom](http://flight-manual.atom.io/hacking-atom/sections/debugging/#update-to-the-latest-version)?
- [x] Did you check the [debugging guide](flight-manual.atom.io/hacking-atom/sections/debugging/)?
- [x] Did you check the [FAQs on Discuss](https://discuss.atom.io/c/faq)?
- [x] Are you reporting to the [correct repository](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#atom-and-packages)?
- [x] Did you [perform a cursory search](https://github.com/issues?q=is%3Aissue+user%3Aatom+-repo%3Aatom%2Felectron) to see if your bug or enhancement is already reported?

For more information on how to write a good [bug report](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#how-do-i-submit-a-good-bug-report) or [enhancement request](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#how-do-i-submit-a-good-enhancement-suggestion), see the `CONTRIBUTING` guide.
### Description

Open multiple windows will lost unsaved modification.
### Steps to Reproduce

1.Open several Atom Windows.
2.Open a Folder.
3.Write Something and don't save in that window.
4.Close the window that you wrote something.
5.Close the rest of Windows.
6.Open Atom and you will found that the everything has gone.
I found that If you open that folder in the new window, the code will be lost.

Normally,if you open one window,they won't be lost.

and if you open another folder first,then open that folder,everything will come back.

**Expected behavior:** [What you expected to happen]
They should come back.
**Actual behavior:** [What actually happened]
They has gone.
### Versions

1.6.0
You can get this information from executing `atom --version` and `apm --version` at the command line.

### LINKED ISSUE #13861: Atom closes without prompting to save
Hello,
I am using Atom 1.14.3 in Mint-17 64-bit and the following scenario causes Atom to close without prompting to save the modified file.

Core setting is:
"Restore previous windows on start" -- UNCHECKED
"Open Empty editor on start" -- CHECKED

Scenario 1 - With just one Untitled tab (as expected)
1) Launch Atom editor from Menu
2) In the Untitled tab (document), type something to make it into a modified but DO NOT save the document
3) Close Atom and it will prompt you to save or to discard the Untitled document.

Scenario 2 - With one Untitled document + One unmodified file (The bug scenario)
1) Launch Atom editor from Menu
2) Open a document but do not modify it
3) If TreeView is shown, toggle it off
4) In the Untitled document, type something to modify it but do not save it. It does not matter how many "Untitled" documents are opened. 
5) Press Ctrl-Q or hit the X to close Atom. Atom does not prompt the user to save the Untitled document and simply discarding it and exit. OOPS! This is difference from Scenario 1.

I have been using an unmodified version of Atom other than to choose the themes from standard ones.

Rover

### LINKED ISSUE #13873: Data loss scenarios with opening project with atom in empty state
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * [x] Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * [x] Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * [x] Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * [x] Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom 
        * There's a lot of issues filed for this already :) this is a issue to consolidate the root issues.  
    * [x] Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

This is to capture a few data loss scenarios with opening your project.  

There are 3 bugs that we can see so far that will cause data loss

### **Opening a Empty Document and try to `Open` the folder from context menu.**

  Repro:

  1. Open Atom, go to settings, check `Open Empty Editor On Start` and uncheck `Restore Previous Windows on Start` 
  2. Open a folder, add a few untitled documents or make changes to file and don't save anything.
  3. Close Atom
  4. Open Atom, you should get a blank editor
  5. Go to context menu and click `Open` (or cmd+o on mac) and open the folder from step 2.  

Expected: Restores Session
Actual: Unsaved session is lost.

### **Opening a Empty Document and trying to `Add a Project Folder` from context menu**

  Repro:

  1. Open Atom, go to settings, check `Open Empty Editor On Start` and uncheck `Restore Previous Windows on Start` 
  2. Open a folder, add a few untitled documents or make changes to file and don't save anything.
  3. Close Atom
  4. Open Atom, you should get a blank editor
  5. Go to context menu and click `Add a Project Folder...` (or cmd+shift+o on mac) and open the folder from step 2.  

Expected: Since there isn't any folder added yet, Atom should restore the session of that folder.
Actual: Unsaved session is lost. 

    
### **Open `atom .` when there is a empty atom editor open.**
Repro:
  1. Open Atom, go to settings, check `Open Empty Editor On Start` and uncheck `Restore Previous Windows on Start` 
  2. Open a folder, add a few untitled documents or make changes to file and don't save anything.
  3. Close Atom
  4. Open Atom, you should get a blank editor
  5. Go to terminal/powershell and go to the folder in step 2 and run `atom .` 

Expected: Restores Session
Actual: Unsaved session is lost. If no atom instance is running this works.

**Reproduces how often:** [What percentage of the time does it reproduce?] 100% of the time.  

### Versions
Atom 1.14.3

### Additional Information

This is causing a lot of various reports, but I think this is the underlying cause for most of them.  

/cc @lee-dohm @iolsen @ben3eee

### PR DESCRIPTION
**Pull request by BinaryMuse**
This PR addresses https://github.com/atom/atom/issues/13873, where the core issue is that re-opening a project when you have a window open with no project folders simply adds those project folders to the existing window, but adding project folders does not restore state for that set of project folders

The primary bad UX that results from this is that re-opening a project from the "Recent Projects" menu when you have a fresh Atom window doesn't really restore the previous project state; it just adds those folders to the project and any unsaved work that's stored in the project state is lost, making the user a [sad panda](http://i.imgur.com/Ddd4lWp.jpg).

This addresses the problem by simply restoring the project state for a folder or set of folders when:

1. The folders are added via "Add Project Folders," or the folders are added automatically via the "Open" menu or via the `atom` command,
2. The folders are being added to an Atom window in which there are zero project folders

This results in [happier pandas](http://i.imgur.com/xlgcLBy.gif).

**Edge Cases**

✅  The Atom window has no project folders but already contains pane items

In this case, we serialize all the current pane items, restore the saved state from the snapshot, and then re-add the serialized pane items to the new workspace. We skip unchanged, empty editors (ones with no changes and with no content, like the one you get when "Open Empty Editor On Start" is enabled), so in the case that the user opens a new window and immediately re-opens a previously open project, the right thing should Just Happen™. It's automagical! 🌈 

restore-project-state-with-open-items

TODO:

- [ ] Don't ask the user to save or discard the changes
- [ ] Make the currently active pane item the active pane item after the state is restored (unless the user specifically opened a file, in which case that file will be active post-restore)

❌  The Atom window has pane items that "conflict" with ones in the state snapshot

If the user already has a file open with unsaved changes, and that same file is open in the saved state snapshot, then the result of the operation is that there will be two copies of the file open, each backed by a different buffer.

This is super edge-casey as it's kinda hard to get into this state (since opening a file adds its parent folder to the project automatically), so I'm not sure if it's one we need to address. 💭 s appreciated.

TODO:

- [ ] ???

❌  Using `project.setPaths` or `project.addPath` does not trigger this behavior

Right now, the code paths for "Application: Add Project Folder" and for opening a file with the "Open" command or with `atom` go through code path that `project.setPaths` and `project.addPath` do not. We opted not to change that behavior since it would be a breaking change.

**Design Considerations**

It turns out that serializing a text editor and then deserializing it into a new `AtomEnvironment` is not very straightforward. The sorta hacky solution implemented here so far is to:

1. Serialize all the current pane items (text editors)
2. Also serialize all the buffers in `@project.buffers`
3. Restore the serialized `AtomEnvironment` state (minus window size and fullscreen setting)
4. Push the serialized buffers from step 2 into the new `@project.buffers`
5. Deserialize the serialized pane items from step 1

This feels 🙅 pretty gross 🙅 but we're not quite sure of a better way to handle it. 💭 s from @atom/core highly appreciated!

---

Fixes #13873
Fixes #13861
Fixes #11293
Fixes #10044

### CODE REVIEW

**Comment by BinaryMuse** [comment]
We caught more interesting edge cases.

1. Have a project open with one folder, and some pane items open.
2. Remove the project folder.
3. Re-add the project folder. The tabs you currently have open will be duplicated (because we add them to the deserialized saved state).

If you repeat steps 2 and 3 again, you'll get an exception: `Uncaught (in promise) Error: Assertion failed: defaultMarkerLayer destroyed at an unexpected time(…)`

**Comment by BinaryMuse** [comment]
Another option is to do away with the whole idea of trying to "merge" the current session with the saved session.

There are four ways in which the user might end up adding a project folder to a window that doesn't have any:

1. `atom .` or `atom /path/to/folder`
2. File -> Open
3. File -> Reopen Project
4. Add Project Folder

In the case of 1, 2, and 3, it seems like we could detect that the current window is "dirty" (has pane items) and then simply open a new Atom window instead of reusing the current one.

In the case of 4, we could notify the user that there is a saved project state, and give them the option of opening that state in a new Atom window, or adding the folder to the current window (which would cause the user to lose that saved state, including any unsaved buffers in that state snapshot).

**Comment by BinaryMuse** [comment]
So we did away with merging the current session as mentioned in the previous comment. 

@ungb would you mind doing some testing to help ensure we covered all the edge cases? 🙏  
Below is a decision tree of the expected behavior after the changes we made:

Note: "clean window" is defined as 1) having an empty project and 2) having no pane items or only empty unnamed buffers

When project is empty and there is saved state associated with the opened/added folders...
* Open a file or folder (from command line or Open menu)
  * If we have a clean window, restore project state in window
  * If window is dirty, restore saved state in new window

* Add folder(s)
  * If we have a clean window, restore project state in window
  * If window is dirty, prompt user to
    * add folder to the existing window LOSING state
    * OR open project folder in a new window

**Comment by ungb** [comment]
LGTM!  I spend some time testing this on Friday.  Only edge case I saw was on mac when there's no windows open and you go to context menu nothing works but that already exist previously.  This is a huge improvement for the data loss scenarios.

**Comment by maxbrunsfeld** [comment]
We've been seeing some intermittent main process test failures on CI that may be caused by this PR. Here's one:

```
  1) AtomApplication launch reuses existing windows when opening paths, but not directories:
     Error: async(5000ms): timed out waiting until window:locations-opened is emitted
      at Promise.resolve.catch.then.result (node_modules/test-until/index.js:56:25)
```

https://circleci.com/gh/atom/atom/3256?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

File: spec/async-spec-helpers.js [hunk 1]
```diff
@@ -60,3 +62,9 @@ function waitsForPromise (fn) {
     })
   })
 }
+
+export function emitterEventPromise (emitter, event, timeout = 5000) {
+  let emitted = false
+  emitter.once(event, () => { emitted = true })
+  return until(`${event} is emitted`, () => emitted, timeout)
+}
```
Comment (maxbrunsfeld) [inline spec/async-spec-helpers.js]: I think we could just resolve this promise explicitly when the event is emitted. The current code works but the polling seems unnecessary.

```js
export function emitterEventPromise (emitter, event, timeout = 5000) {
  return new Promise((resolve, reject) => {
    const timeoutHandle = setTimeout(reject, timeout)
    emitter.once(event, () => {
      clearTimeout(timeoutHandle)
      resolve()
    })
  })
}
```

**Pull request merged**